pkg/io/buffered.go

```diff
@@ -0,0 +1,283 @@
+// Subsequent reads are returned from the buffers. Additionally it supports concurrent readers
+// by maintaining multiple buffers at different offsets, and matching up reads with existing
+// buffers where possible. When needed the least-recently-used buffer is overwritten with new reads.
+type BufferedReaderAt struct {
+	mtx     sync.Mutex
+	ra      io.ReaderAt
+	rasz    int64
+	rdsz    int
+	count   int64
+	buffers []readerBuffer
+}
+
+type readerBuffer struct {
+	mtx   sync.RWMutex
+	buf   []byte
+	off   int64
+	count int64
+}
+
+var _ io.ReaderAt = (*BufferedReaderAt)(nil)
+
+func NewBufferedReaderAt(ra io.ReaderAt, readerSize int64, bufSize, bufCount int) *BufferedReaderAt {
+	r := &BufferedReaderAt{
+		ra:      ra,
+		rasz:    readerSize,
+		rdsz:    bufSize,
+		buffers: make([]readerBuffer, bufCount),
+	}
+
+	return r
+}
+
+func (r *BufferedReaderAt) canRead(buf *readerBuffer, offset, length int64) bool {
+	return offset >= buf.off && (offset+length <= buf.off+int64(len(buf.buf)))
+}
+
+func (r *BufferedReaderAt) read(buf *readerBuffer, b []byte, offset int64) {
+	start := offset - buf.off
+	copy(b, buf.buf[start:start+int64(len(b))])
+}
+
+func (r *BufferedReaderAt) prep(buf *readerBuffer, offset, length int64) {
+	offset, sz := calculateBounds(offset, length, r.rdsz, r.rasz)
+
+	// Realloc?
+	if int64(cap(buf.buf)) < sz {
+		buf.buf = make([]byte, sz)
+	}
+	buf.buf = buf.buf[:sz]
+	buf.off = offset
+}
+
+func (r *BufferedReaderAt) populate(buf *readerBuffer) (int, error) {
+	// Read
+	n, err := r.ra.ReadAt(buf.buf, buf.off)
+	return n, err
+}
+
+func calculateBounds(offset, length int64, bufferSize int, readerAtSize int64) (newOffset, newLength int64) {
+	// Increase to minimim read size
+	sz := length
+	if sz < int64(bufferSize) {
+		sz = int64(bufferSize)
+	}
+
+	// Don't read larger than entire contents
+	if sz > readerAtSize {
+		sz = readerAtSize
+	}
+
+	// If read extends past the end of reader,
+	// back offset up to fill the whole buffer
+	if offset+sz >= readerAtSize {
+		offset = readerAtSize - sz
+	}
+
+	return offset, sz
+}
+
+func (r *BufferedReaderAt) ReadAt(b []byte, offset int64) (int, error) {
+	// There are two-levels of locking: the top-level governs the
+	// the reader and the arrangement and position of the buffers.
+	// Then each individual buffer has its own lock for populating
+	// and reading it.
+
+	// The main reason for this is to support concurrent activity
+	// while solving the stampeding herd issue for fresh reads:
+	// The first read will prep the offset/length of the buffer
+	// and then switch to the buffer's write-lock while populating it.
+	// The second read will inspect the offset/length and know
+	// that it will satisfy, but by taking the read-lock will
+	// wait until the first call has finished populating the buffer .
+
+	r.mtx.Lock()
+
+	if len(r.buffers) == 0 {
+		r.mtx.Unlock()
+		return r.ra.ReadAt(b, offset)
+	}
+
+	// Least-recently-used tracking
+	r.count++
```

Comment:
We can only do 292,471,208,678 reads/second for a year before we trip this bug. :)

Reply:
:) Yes the extreme simpleness of this approach was considered, but the readerAt is used once per request, so it's fine.